### ISSUE TITLE
Server should advertise that posts are local-only so third-party clients can theoretically support them

### ISSUE THREAD

**Issue by dariusk** [hometown-fork/hometown#23]
Local-only posting is not well supported on mobile apps.

The problem is that the local-only posting has no way of "advertising" that it exists as a feature. Or if it does, it's not really documented! This bug will track my ongoing work on this issue.

**Comment by dariusk** [comment]
Mastodon v3.0.0rc1 is introducing a "server Actor" which may be a good place to include this sort of thing. A client could perhaps query the server actor to see what features it supports.

**Comment by marrus-sh** [comment]
For the record, `/api/v1/instance` has historically been the place that people put this sort of information, if the only concern is Mastodon API apps. (Since local-only posting doesn't federate, I would think it should be).

As far as I understand (I may not be well-informed), the server Actor is more for AP federation, and you *could* advertise local-only posting there as well, but it would not show up to apps (as they do not use the AP model).

(For convenience, I believe https://github.com/hometown-fork/hometown/blob/hometown-dev/app/serializers/rest/instance_serializer.rb is the file used to control the contents of `/api/v1/instance`.)

**Comment by dariusk** [comment]
Thanks, that might make more sense. I'm struggling to see an example on a live server but I suppose `registrations: "false"` is an example of feature advertisement.

**Comment by marrus-sh** [comment]
the GlitchSoc fork (e.g., <https://dev.glitch.social/api/v1/instance>) has a bit more, in the form of `max_toot_chars` and `poll_limits`

but yeah it is definitely the case that most people who write extensions don't bother adding a property to advertise it, and upstream just assumes that all apps will know the given featureset from the version string.

however, it is why the endpoint was originally created: https://github.com/tootsuite/mastodon/pull/654 :P

(edit, because I just realized it wasn't obvious: that PR introduces `/api/v1/site`, it was renamed to `/api/v1/instance` here: https://github.com/tootsuite/mastodon/commit/6be7bde24378bcb034552248021ea2450bb10524

(edit edit: wow i just realized that's when eugen decided to drop `max_toot_chars` from upstream; i knew it didn't last long :P. his argument has always been "mastodon is like email; clients offer their own features but all servers use the same api", which is why he has historically opposed advertising specific features through the API. i have always thought this was a bad argument :P.))

**Comment by marrus-sh** [comment]
(fun historical note, if you weren't aware; when GlitchSoc originally implemented local-only posts, we used typing an 👁 emoji at the end of a post to toggle local-only mode. the reason for this hack was precisely because it meant that it would be supported across all apps.

not saying you should do this, but it might be why we never got around to actually advertising the capability anywhere lol. however, i'm sure glitch will pick up whatever you decide and start advertising in a similar manner as well.)

**Comment by dariusk** [comment]
Yeah I actually really respect the eyeball emoji hack, it's a great idea!

**Comment by Aldarone** [comment]
It's a bit wider than « local only » posts but I just opened the issue glitch-soc/mastodon#1592 to at least expose the fork's name in the `instance` endpoint of the api so it'd look like that : 

- Currently : `3.4.0`
- Proposal : `3.4.0 (compatible; hometown 1.0.4)`

Obviously it's not a silver bullet because client would need to be aware of each fork but it's still better than nothing and they'd be able to support the most popular of them.

**Comment by rscmbbng** [comment]
There is also some discussion on the matter here: https://github.com/glitch-soc/mastodon/pull/1438 and here https://github.com/tuskyapp/Tusky/issues/835 

Seems like clients are waiting for servers to define a standard vocabulary. How could we move this conversation forwards?

**Comment by igalic** [comment]
there's a configuration API now: https://github.com/mastodon/mastodon/pull/16485
